### PR TITLE
fix:  rendezvous identify example

### LINKED ISSUE #5388: rendezvous: Example "discovering with identify" doesn't work
### Summary

When launching the [`rendezvous` example](https://github.com/libp2p/rust-libp2p/tree/master/examples/rendezvous) and trying to discover peers with `identify`, the peer abort because it can't register without knowing any external address.
**Steps to reproduce**
1. [Launch the server (1)](https://github.com/libp2p/rust-libp2p/tree/master/examples/rendezvous#usage)
2. [Launch a peer with identify (4)](https://github.com/libp2p/rust-libp2p/tree/master/examples/rendezvous#usage)

### Expected behavior

As suggested in [this comment](https://github.com/libp2p/rust-libp2p/blob/master/examples/rendezvous/src/bin/rzv-identify.rs#L77), the protocol `identify` could allow to know and set an external address automatically and then the registration should be successfull. 
Then the discovering would work as described.

### Actual behavior

The peer aborts when registering because it doesn't know any external address.

### Relevant log output

```shell
$ RUST_LOG=info cargo run --bin rzv-identify
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.15s
     Running `~/rust-libp2p/target/debug/rzv-identify`
2024-05-15T06:34:58.587487Z  INFO libp2p_swarm: local_peer_id=12D3KooWNFn1BxcTXDvNY9YKY3gr4Sn3V6eTu2Pb1dPnFgtHdTHF
2024-05-15T06:34:58.587938Z  INFO rzv_identify: Listening on /ip4/127.0.0.1/tcp/37627
2024-05-15T06:34:58.587976Z  INFO rzv_identify: Listening on /ip4/192.168.1.59/tcp/37627
2024-05-15T06:34:58.675236Z ERROR rzv_identify: Failed to register: We don't know about any externally reachable addresses of ours
```

### Possible Solution

As suggested in [this post](https://discuss.libp2p.io/t/how-we-discover-our-own-diallable-ip-address-and-port/219), we could use the [observed address](https://docs.rs/libp2p/latest/libp2p/identify/struct.Info.html#structfield.observed_addr) as a potential `ExternalAddress`. 
Then we could either use it directly or use `autonat` to probe it and obtain a `ConfirmedExternalAddress`

### Version

libp2p version `0.53.2`

### Would you like to work on fixing this bug ?

Yes

### PR DESCRIPTION
**Pull request by tsnewnami**
## Description
This pr closes https://github.com/libp2p/rust-libp2p/issues/5388 by explicitly adding the local observed address, noting this is out of protocol behaviour in non-example cases

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Comment by tsnewnami** [comment]
Hey @jxs, is there a `CHANGELOG.md` edit needed for this? 
Thank you.

File: examples/rendezvous/src/bin/rzv-identify.rs [hunk 1]
```diff
@@ -76,8 +76,13 @@ async fn main() {
             }
             // once `/identify` did its job, we know our external address and can register
             SwarmEvent::Behaviour(MyBehaviourEvent::Identify(identify::Event::Received {
+                info,
                 ..
             })) => {
+                // register our external address. this is not normal behaviour under identify, usually
+                // local addresses need to be added manually. this is done for the sake of
+                // running the example.
```
Comment (jxs) [inline examples/rendezvous/src/bin/rzv-identify.rs]:
```suggestion
                // Register our external address. Needs to be done explicitly 
                // for this case, as it's a local address.
```

**Review by jxs — DISMISSED** [review]

File: examples/rendezvous/src/bin/rzv-identify.rs [hunk 2]
```diff
@@ -76,8 +76,12 @@ async fn main() {
             }
             // once `/identify` did its job, we know our external address and can register
             SwarmEvent::Behaviour(MyBehaviourEvent::Identify(identify::Event::Received {
+                info,
                 ..
             })) => {
+                // Register our external address. Needs to be done explicitly
```
Comment (jxs) [inline examples/rendezvous/src/bin/rzv-identify.rs]:
```suggestion
                // Register our external address. Needs to be done explicitly
```

**Comment by tsnewnami** [comment]
Thanks for fixing that up @jxs. Happy to merge 🙏

**Review by jxs — APPROVED** [review]

**Pull request merged**